internal/daemon/metric/instrument_handlers.go

```diff
@@ -66,6 +69,87 @@ func (sh *statsHandler) HandleRPC(ctx context.Context, s stats.RPCStats) {
 	}
 }
 
+type requestRecorder struct {
+	reqLatency prometheus.ObserverVec
+	labels     prometheus.Labels
+
+	// measurements
+	start time.Time
+}
+
+// NewGrpcRequestRecorder creates a requestRecorder struct which is used to measure gRPC client request latencies.
+func NewGrpcRequestRecorder(fullMethodName string, reqLatency prometheus.ObserverVec) requestRecorder {
+	service, method := SplitMethodName(fullMethodName)
+	r := requestRecorder{
+		reqLatency: reqLatency,
+		labels: prometheus.Labels{
+			LabelGrpcMethod:  method,
+			LabelGrpcService: service,
+		},
+		start: time.Now(),
+	}
+
+	return r
+}
+
+func (r requestRecorder) Record(err error) {
+	r.labels[LabelGrpcCode] = StatusFromError(err).Code().String()
+	r.reqLatency.With(r.labels).Observe(time.Since(r.start).Seconds())
+}
+
+type connectionTrackingListener struct {
+	net.Listener
+	acceptedConns      prometheus.Counter
+	closedConnsSuccess prometheus.Counter
+	closedConnsError   prometheus.Counter
+}
+
+// NewConnectionTrackingListener registers a new Prometheus gauge with an unique
+// connection type label and wraps an existing listener to track when connections
+// are accepted and closed.
+// Multiple calls to Close() a listener connection will only decrement the gauge
+// once. A call to Close() will decrement the gauge even if Close() errors.
+func NewConnectionTrackingListener(l net.Listener, purpose string, ac prometheus.CounterVec, cc prometheus.CounterVec) *connectionTrackingListener {
+	p := prometheus.Labels{LabelConnectionPurpose: purpose}
+	pSuccess := prometheus.Labels{LabelConnectionPurpose: purpose, LabelDisconnectionStatus: "success"}
+	pError := prometheus.Labels{LabelConnectionPurpose: purpose, LabelDisconnectionStatus: "error"}
```

Comment:
That's true. For me it was a tossup between feeling weird passing it through here or feeling weird populating the labels at close time. I do like not having to pass two counters for closing though.

If I do it the latter way, I don't know how to avoid the problem of giving the listener wrapper a counter for the accepted connections and a curried counter vec (which is later given a status label value) for the closed connections. This makes things a bit awkward for the testing function, but in hindsight is still better than having to make two counters to track closed connections